compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/PromisedValue.kt

```diff
@@ -59,7 +60,17 @@ abstract class PromisedValue(val codegen: ExpressionCodegen, val type: Type, val
         }
 
         if (type != target || (castForReified && irType.anyTypeArgument { it.isReified })) {
-            StackValue.coerce(type, target, mv, type == target)
+            if (codegen.typeMapper.mapType(erasedSourceType) == type && codegen.typeMapper.mapType(erasedTargetType) == target) {
+                // what we really want to know here is if `type` is subType of `target` and if `target` is from Java. Before we find
+                // a better way to do this, in order to avoid introduce unexpected breaking changes, it is important to determine whether
+                // `erasedSourceType` and `erasedTargetType` are appropriate for the calculation of `isSubType` and `isFromJava`
+                val isSubType = erasedSourceType.isSubtypeOfClass(erasedTargetType.classOrNull!!)
+                val isFromJava = (erasedTargetType.classifierOrFail.owner as IrDeclaration).isFromJava()
+
+                // we might need to cast to the correct upper bound when irType is multiple upper bounds type parameter
+                val coerceTarget = computeCoerceTarget(isSubType, erasedTargetType, default = target)
+                StackValue.coerce(type, coerceTarget, mv, type == coerceTarget, isSubType && isFromJava)
+            } else StackValue.coerce(type, target, mv, type == target, false)
```

Comment:
@dnpetrov  Sorry about the mistake. I should be more careful here.

the problem is `erasedSourceType.isSubtypeOfClass(erasedTargetType.classOrNull!!)` doesn't always mean `type` is subType of `target` 

As the comment describled, I have taken a more cautious approach here,  if we can't rely on `erasedSourceType` and `erasedTargetType`, I think we should fall back to original logic then.